src/Adapter/State/QueryHandler/GetStateWithIsoCodeExistsHandler.php

```diff
@@ -0,0 +1,56 @@
+<?php
+/**
+ * 2007-2019 PrestaShop and Contributors
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Adapter\State\QueryHandler;
+
+use PrestaShop\PrestaShop\Core\Domain\State\Query\GetStateWithIsoCodeExists;
+use PrestaShop\PrestaShop\Core\Domain\State\QueryHandler\GetStateWithIsoCodeExistsHandlerInterface;
+use State;
+
+/**
+ * Handles query for determining if state with given iso code exists
+ */
+final class GetStateWithIsoCodeExistsHandler implements GetStateWithIsoCodeExistsHandlerInterface
```

Comment:
Not sure about this, maybe you've discussed it with @matks? I think we have used some sort of providers in adapters. Ping just to make sure. 
Also, @matks, should we prefix all queries with get, even if its returning bool. or maybe there is any other convention? Cuz now its hard to read the name.

Reply:
Ah, crap, again this debate 😅 

We brainstormed a lot with @RaimondasSapola Raimondas on this one but we finally figured it out:

We should use Queries and Commands to model interactions between the User and the Domain.
In a CRUD add/edit form, the Commands AddObjectCommand and UpdateObjectCommand actually carry both the data and the intention of the user to the Domain.

Here, the GetStateWithIsoCodeExistsHandler is used by a constraint to validate the form. There is no user in this interaction, so CQRS is invalid.

Indeed what we should use instead here is a Read Model (=> an Adapter DataProvider) which is another concept linked to CQRS.

Commands and Queries are user-interactions modeled into data. If the user is interested to know if a state with ISO code "TAH" exists, then we can create such Query. If it's not a user need but rather a system need, like here, then no Query.

I'm sorry this is quite abstract 😢 please tell me if this needs further explanations.

Reply:
All in all, it should be some kind of data provider instead, as its not a real user use-case, but rather a system required check.

Reply:
very informative explanation!

Reply:
So will refactor GetStateWithIsoCodeExistsHandler to => CountryStateByIsoCodeChoiceProvider

Reply:
done :building_construction: